google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITE2ETracingTest.java

```diff
@@ -0,0 +1,1251 @@
+// (3).
+// 5. Traces are read-back using TraceServiceClient and verified against expected Call Stacks.
+// TODO In the future it would be great to have a single test-driver for this test and
+// ITTracingTest.
+@RunWith(TestParameterInjector.class)
+public class ITE2ETracingTest extends ITBaseTest {
+
+  protected boolean isUsingGlobalOpenTelemetrySDK() {
+    return useGlobalOpenTelemetrySDK;
+  }
+
+  // Helper class to track call-stacks in a trace
+  protected class TraceContainer {
+
+    // Maps Span ID to TraceSpan
+    private Map<Long, TraceSpan> idSpanMap;
+
+    // Maps Parent Span ID to a list of Child SpanIDs, useful for top-down traversal
+    private Map<Long, List<Long>> parentChildIdMap;
+
+    // Tracks the Root Span ID
+    private long rootId;
+
+    public TraceContainer(String rootSpanName, Trace trace) {
+      idSpanMap = new TreeMap<Long, TraceSpan>();
+      parentChildIdMap = new TreeMap<Long, List<Long>>();
+      for (TraceSpan span : trace.getSpansList()) {
+        long spanId = span.getSpanId();
+        idSpanMap.put(spanId, span);
+        if (rootSpanName.equals(span.getName())) {
+          rootId = span.getSpanId();
+        }
+
+        // Add self as a child of the parent span
+        if (!parentChildIdMap.containsKey(span.getParentSpanId())) {
+          parentChildIdMap.put(span.getParentSpanId(), new ArrayList<Long>());
+        }
+        parentChildIdMap.get(span.getParentSpanId()).add(spanId);
+      }
+    }
+
+    String spanName(long spanId) {
+      return idSpanMap.get(spanId).getName();
+    }
+
+    List<Long> childSpans(long spanId) {
+      return parentChildIdMap.get(spanId);
+    }
+
+    // This method only works for matching call stacks with traces which have children of distinct
+    // type at all
+    // levels. This is good enough as the intention is to validate if the e2e path is WAI - the
+    // intention is not to validate Cloud Trace's correctness w.r.t. durability of all kinds of
+    // traces.
+    boolean containsCallStack(String... callStack) throws RuntimeException {
+      ArrayList<String> expectedCallStack = new ArrayList<String>();
+      for (String call : callStack) {
+        expectedCallStack.add(call);
+      }
+      if (expectedCallStack.isEmpty()) {
+        throw new RuntimeException("Input callStack is empty");
+      }
+      return dfsContainsCallStack(rootId, expectedCallStack);
+    }
+
+    // Depth-first check for call stack in the trace
+    private boolean dfsContainsCallStack(long spanId, List<String> expectedCallStack) {
+      logger.info(
+          "span="
+              + spanName(spanId)
+              + ", expectedCallStack[0]="
+              + (expectedCallStack.isEmpty() ? "null" : expectedCallStack.get(0)));
+      if (!expectedCallStack.isEmpty() && spanName(spanId).equals(expectedCallStack.get(0))) {
+        if (childSpans(spanId) == null) {
```

Comment:
I also think it'd be better for `childSpans()` function to return an empty list rather than null for cases where there is no more children. Once it does that, it'd be nice to get rid of this if-else clause to further simplify this function.

Reply:
Don't we still need to check if the `childSpans(spanId).isEmpty()` - that's the recursion termination condition so there's no way around that.

Reply:
oops. my mistake.